flang/lib/Optimizer/Builder/FIRBuilder.cpp

```diff
@@ -279,6 +279,9 @@ mlir::Block *fir::FirOpBuilder::getAllocaBlock() {
   if (auto cufKernelOp = getRegion().getParentOfType<cuf::KernelOp>())
     return &cufKernelOp.getRegion().front();
 
+  if (auto doConcurentOp = getRegion().getParentOfType<fir::DoConcurrentOp>())
+    return doConcurentOp.getBody();
```

Comment:
Do we need this for any allocation in the do concurrent? We did not have this so far so I'm just wondering.

Reply:
We do this way for the new `fir.do_concurrent` and `fir.do_concurrent.loop` ops to keep all info related to the loop nest closely located in the IR. The allocations needed here are for the nest iteration variables. See: https://discourse.llvm.org/t/modeling-do-concurrent-loops-in-the-fir-dialect/84950/6?u=bob.belcher and related discussion and https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Dialect/FIROps.td#L3457.

Reply:
Ok for the iteration variables but what about any temporary that will be allocated during the body lowering. Is it ok to add them at the same place as the iteration variables?

Reply:
These will be emitted in the body of the loop itself (for every iteration), just like what would happen now when targeting the old `fir.do_loop ... unorderd` nest. I think it would be ok to move them to the alloca block only if we serialize to `fir.do_loop ... unordered` nest (as we do in `DoConcurrentConversion` in `SimplifyFIROperations.cpp`) or if we are targeting OpenMP, for example, we can share them for each chunk of the parallel loop that we map to. But in the general case, I think it would not be ok to move them out at this point.

Let me know if I misunderstood what you meant.

Reply:
I don't think they are created in the loop body. The code calls `getAllocaBlock()` and currently, it will probably return the entry block of the function. With your change it will return the do concurrent body. So instead of having the alloca for temporary in the function entry block you will have them in the do concurrent. If this is fine for you then it's ok but this change from the current behavior.

Reply:
Ah you are right! I have a PR that handles this situation for OpenMP: https://github.com/llvm/llvm-project/pull/127635. So I think it should be fine at least for now.

Reply:
You are right, I got confused here. This is still fine since the temp allocations will be emitted in the `fir.do_concurrent` wrapper op not the `fir.do_concurrent.loop` op. When serialized (i.e. converted to `fir.do_loop ... unordered` nest), these allocations are moved to the alloca block (see https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/Transforms/SimplifyFIROperations.cpp#L167-L176). For OpenMP, I call these "loop-local values", these values are privatized by the do concurrent to OpenMP pass (see https://github.com/llvm/llvm-project/pull/127635).